### ISSUE TITLE
Feature request: new value for the preserveParens option

### ISSUE THREAD

**Issue by thorn0** [meriyah/meriyah#183]
Would be good to have a mode in which `ParenthesizedExpression` nodes would be created only for parentheses used for JSDoc type assertions.

The syntax of JSDoc type assertions is simple: `@type {...}` JSDoc comment followed by a parenthesized expression. Anything else is not a type assertion. Reference: [Closure Compiler](https://github.com/google/closure-compiler/wiki/Annotating-JavaScript-for-the-Closure-Compiler#type-casting), [TypeScript](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#casts).

**Comment by 3cp** [comment]
This is probably difficult for meriyah, as the content of comment needs to go through some kind of syntax parsing too.

**Comment by thorn0** [comment]
The syntax is super simple. See the links. No need to parse the type itself. Just the fact that the comment starts with `*` and contains `@type` followed by at least something.

**Comment by 3cp** [comment]
I see.

**Comment by 3cp** [comment]
@KFlash can you have a look, I am not sure how to temporarily turn on Context.OptionsPreserveParens if next token is `(`. Also has to properly reset it after the ParenthesizedExpression. I sensed it might be complicated if need to support recursion .

I am not sure if recursive type casting is possible. @thorn0 ?
```js
/** @type xxx */ ({a: /** @type xxx */ (b) })
```

Right now, I guess you are turning on preserveParens option globally.

**Comment by KFlash** [comment]
Can you just create a new options? Name i `thorn`. Then you do the same as for other options

`if (options.thorn) context | Context.OptionsThorn |  OptionsPreserveParens `

**Comment by thorn0** [comment]
> I am not sure if recursive type casting is possible

Yes, possible. Also they can be nested directly (people really use this):
```js
/** @type MyType */ (/** @type any */ (x))
```

> What's the use case for this? And why do you want to turn off ParenthesizedExpression ? It's in the spec 

If you mean ESTree, `ParenthesizedExpression` is not there. It's a non-standard extension. Overall, ESTree's approach is to ignore insignificant parentheses. However, the JSDoc convention that uses such parentheses is widespread and important. So from the point of view of tools like formatters and linters, it makes sense to create these non-standard nodes only for JSDoc. In other cases, they only get in the way. E.g. Prettier doesn't simply print the AST. It also does a lot of checks of the AST structure along the way to choose between different ways to print things. Taking `ParenthesizedExpression` nodes into account in all those checks would ridiculously complicate them, so actually what happens now in Prettier is an extra step that removes all `ParenthesizedExpression`s that aren't JSDoc type assertions from the AST before passing it to printing.

**Comment by fisker** [comment]
I'm going to close this as not planed since the current `preserveParens` already works for Prettier https://github.com/prettier/prettier/pull/17600 .
